vision/use-cases/hey_llm/src/main.ts

```diff
@@ -0,0 +1,506 @@
+  const prompt = `
+## Instruction
+${instruction}
+
+## Context (CSV formatted)
+${context ? context.map(row => row.join(', ')).join('\n') : ''}
+
+## Task
+Input: ${input}
+Output:`;
+
+  const cache = CacheService.getDocumentCache();
+  const cacheKey = `hey_llm:${generateHashValue(prompt)}`;
+  const cached = cache?.get(cacheKey);
+  if (cached) return cached;
+
+  const oauthService = getGoogleService_();
+  const url = `https://${LOCATION}-aiplatform.googleapis.com/v1/projects/${getProjectNumber_()}/locations/${LOCATION}/publishers/google/models/${GEMINI_MODEL}:generateContent`;
+
+  const payload = JSON.stringify({
+    contents: [
+      {
+        role: 'user',
+        parts: [
+          {
+            text: prompt,
+          },
+        ],
+      },
+    ],
+    systemInstruction: {
+      parts: [
+        {
+          text: 'The response is to be used in side a spreadsheet cell and needs to be concise. Just show the answer only.',
+        },
+      ],
+    },
+  });
+  const res = UrlFetchApp.fetch(url, {
+    method: 'post',
+    headers: {
+      Authorization: 'Bearer ' + oauthService.getAccessToken(),
+    },
+    contentType: 'application/json',
+    payload: payload,
+    muteHttpExceptions: true,
+  });
+  const result: GenerateContentResponse = JSON.parse(res.getContentText());
+  if (!(result.candidates && result.candidates[0].content.parts[0].text)) {
+    Logger.log(result);
+    throw new Error('Request to Gemini failed. ' + res.getContentText());
+  }
+  const out = result.candidates[0].content.parts[0].text.trim();
+  cache?.put(cacheKey, out);
+  return out;
+}
+
+/**
+ * Removes invalid characters from the file name.
+ * @param filename Source file name
+ * @returns Sanitized file name
+ */
+function sanitizeFileName(filename: string) {
+  return filename.replace(/[/\\?%*:|'"<>]/g, '_');
+}
+
+/**
+ * Gets a thumbnail URL for the image file if available in the parent Drive folder.
+ * @private
+ * @param oauth OAuth2 Service
+ * @param filename Image file name to check
+ * @param parentFolderID Parent Drive folder to check
+ * @returns Thumbnail URL if available
+ */
+function checkDriveImage_(
+  oauth: GoogleAppsScriptOAuth2.OAuth2Service,
+  filename: string,
+  parentFolderID: string,
+) {
+  const q = encodeURI(
+    `name = '${filename}' and '${parentFolderID}' in parents`,
+  );
+  const url = `https://www.googleapis.com/drive/v3/files?q=${q}&fields=files(id,name,thumbnailLink)`;
+  const res = UrlFetchApp.fetch(url, {
+    method: 'get',
+    headers: {
+      Authorization: 'Bearer ' + oauth.getAccessToken(),
+    },
+    muteHttpExceptions: true,
+  });
+  const response: {
+    files?: {id: string; name: string; thumbnailLink: string}[];
+  } = JSON.parse(res.getContentText());
+
+  if (!response.files) {
+    Logger.log(response);
+    throw new Error('no files included in the response');
+  }
+
+  if (response.files.length > 0) {
+    return response.files[0].thumbnailLink;
+  }
+  return undefined;
```

Comment:
Is there a Library you can use for Drive instead of doing raw API requests?

Reply:
With the same reason as the above comment, I need to construct the raw request to comply with AppScript's restriction in making external network calls via `UrlFetchApp`.